Feature add AsyncMoya (#2233)
Moya/Moya: docs/AsyncMoya.md

# Async/await

Moya provides an optional `Async/await` implementation of
`MoyaProvider` that does a few interesting things. Instead of
calling the `request()` method and providing a callback closure
to be executed when the request completes, we use `await`.

```swift
let provider = MoyaProvider<GitHub>()
```

After that simple setup, you're off to the races:

```swift
AsyncTask {

>     let result = await provider.request(.zen) //return type `Result<Response, MoyaError>`

Reviewer: You mean `asyncRequest`?

```suggestion
    let result = await provider.asyncRequest(.zen) //return type `Result<Response, MoyaError>`
```

    switch result {
    case let .success(response):
        // do something with the data
    case let .failure(error):
        // handle the error
    default:
        break
    }
}
```

`AsyncTask` is `typealias` for Apple mechanism `Task` for resolve conflict with `Moya.Task`.

You can also use `requestWithProgress` to track progress of 
your request:
```swift
await provider.requestWithProgress(SimpleTarget.posts).forEach({ result in
    switch result {
    case let .success(progress):
        if let response = progress.response {
            // do something with response
        } else {

>             print("Progress: \(progressResponse.progress)")

Reviewer: Also `progressResponse` is not declared, can you replace it with the correct example?

        }
    case let .failure(error):
        break
    }
})
```

or you can use `for in` loop:

```swift
for await progressResponse in await provider.requestWithProgress(SimpleTarget.posts) {
    case let .success(progress):
        if let response = progress.response {
            // do something with response
        } else {

>             print("Progress: \(progressResponse.progress)")

Reviewer: Same here

Author: oops, just a moment

# Async/await

Moya provides an optional `Async/await` implementation of
`MoyaProvider` that does a few interesting things. Instead of
calling the `request()` method and providing a callback closure
to be executed when the request completes, we use `await`.

```swift
let provider = MoyaProvider<GitHub>()
```

After that simple setup, you're off to the races:

```swift
AsyncTask {
    let result = await provider.asyncRequest(.zen) //return type `Result<Response, MoyaError>`
    switch result {
    case let .success(response):
        // do something with the data
    case let .failure(error):
        // handle the error
    default:

>         break

Reviewer 2: I don't think we need this `default` case as `Result` should be `@frozen`, not accepting any new cases in its enum.

Author: removed

    }
}
```

> `AsyncTask` is `typealias` for Apple mechanism `Task` for resolve conflict with `Moya.Task`.

Reviewer 2: I think it may be worth thinking about considering to rename `Moya.Task`, as this is (?) referring to another construct than `Task` in an async context, and may prove confusing going forward. What do you think?

Author: yes, i think Moya.Task should be renamed, but this will "breaking changes" for all and i haven't idea how to do it better now :(

Reviewer 2: We could at least think of moving away/renaming `Moya.Task`. You could create an issue for that if you wish. :)

Reviewer 2: At least deprecating `Moya.Task` as well may be a good thing to do here, so that in a future (major) release, we can move to the desired naming (and use `Swift.Task` directly).

Author: good idea 👍🏻 
i can mark `Moya.Task` like deprecate in future version, so that users are ready)

Author: done ✅

You can also use `requestWithProgress` to track progress of 
your request:
```swift
AsyncTask {
    await provider.requestWithProgress(SimpleTarget.posts).forEach({ result in
        switch result {
        case let .success(progress):
            if let response = progress.response {
                // do something with response
            } else {

>                 print("Progress: \(progressResponse.progress)")

Reviewer: It's seems you missed this one:

```suggestion
                print("Progress: \(progress.progress)")
```

Author: fixed

            }
        case let .failure(error):
            break
        }
    })
}
```

or you can use `for in` loop:

```swift
AsyncTask {
    for await progressResponse in await provider.requestWithProgress(SimpleTarget.posts) {
        switch progressResponse {
        case let .success(progress):
            if let response = progress.response {
                // do something with response
            } else {
                print("Progress: \(progress.progress)")
            }
        case let .failure(error):
            print(error.localizedDescription)
        }
    }
}
```

> Request with progress use `AsyncStream` mechanism you can use with him all functional operations like `.map`, `.filter`, `.flatMap` e.t.c.

Reviewer 2: ```suggestion
Request with progress use `AsyncStream` mechanism you can use functional operations like `.map`, `.filter`, `.flatMap` etc.
```
